### PR TITLE
Inform users when a `ddev` update is available

### PR DESCRIPTION
**Pull request by beeradb**
## The Problem:
There is a need for users to be notified of updates to ddev. 

## The Fix:
This code introduces periodic (weekly) checking of github releases. Using semver comparisons it informs the user if there are available updates by linking them to the release URL.

Overview:
* Update checks are weekly
* Currently there is no major/minor/patch release versioning or anything along those lines. We may want to add that in the future, but I wanted to keep things simple for now. If there's a newer release, you'll get a weekly message informing you of the release.
* Updates do not run if it's not an official release. i.e. dev builds with -dirty or commitish as the current version do not get updates ran. This is currently a very naive check which just looks for a "-" in the version and doesn't run the update if it exists.
* A file is stored at ~/.ddev/.update. File modification date is used to track time since last update.
* The package which controls this is fairly general. I believe it could be used for any of our projects, as you just give it a repository and current version and it lets you know if an update is available, and what the URL for the update is. 

## The Test:

The easiest way to test is the following:
- Build this binary with an old version. For instance, `make DdevVersion=v0.2 darwin`
- Run any ddev command, such as `ddev version`
- The first run you should see information about an available update. Subsequent runs should not have it (until a week has passed)
- You can alter the file modification date using the touch command. The following command modifies it (on OSX) to May 1st 2017 at 01:01 AM. `touch -mt 201705010101  ~/.ddev/.update`
- After altering the modification time you should see the update message again. An alteration of the file modification date to be less than a week old should not trigger a new update message.

## Automation Overview:
Please see `updatecheck_test.go` in the `updatecheck` package.

## Related Issue Link(s):
The parent issue for this is at https://github.com/drud/ddev/issues/99

### CODE REVIEW

File: pkg/updatecheck/updatecheck_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	assert.True(updateRequired, "Update is not required when the update file does not exist")
+	assert.NoError(err)
+
+	// Ensure updates are not required when the update duration is impossibly far in the future.
+	updateRequired, err = IsUpdateNeeded(updateFile, time.Duration(9999999)*time.Second)
+	assert.False(updateRequired, "Update is not required when the update file does not exist")
+	assert.NoError(err)
+
+	// Sleep for 2 seconds.
+	time.Sleep(time.Duration(2) * time.Second)
```
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Could be wrong, but I don't think this does what you want. time.Duration takes ns right?  Do you want time.Sleep(2 * time.Second) ?  Sad to see a sleep but doesn't seem like it's the end of the world.

File: pkg/updatecheck/updatecheck_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	// Ensure updates are required for a duration lower than the sleep.
+	updateRequired, err = IsUpdateNeeded(updateFile, time.Duration(1)*time.Second)
```
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Again, I don't *think* time.Duration() is what you want here.

File: pkg/updatecheck/updatecheck_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	assert := assert.New(t)
+	var versionTests = []struct {
+		in  string
+		out bool
+	}{
+		{"0.1.0", true},
+		{"v0.1.0", true},
+		{"v19.99.99", true},
+		{"19.99.99-8us8dfgh7-dirty", false},
+		{"v0.3-7-g3ca5586-dirty", false},
```
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Nicely done array of test cases. Highly approved.

File: pkg/updatecheck/updatecheck_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+func TestIsReleaseVersion(t *testing.T) {
+	assert := assert.New(t)
+	var versionTests = []struct {
+		in  string
+		out bool
+	}{
+		{"0.1.0", true},
+		{"v0.1.0", true},
+		{"v19.99.99", true},
+		{"19.99.99-8us8dfgh7-dirty", false},
```
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Not sure how a release with these names like -dirty could be created. But your strategy of failing on a - seems ok. However,  that's really just something we use with git and wouldn't really be a real live thing I don't think.

**Comment by rickmanelius** [comment]
Beginning my test immediately after stand-up.

File: pkg/updatecheck/updatecheck_test.go (hunk 4 above)
Comment (beeradb) [inline pkg/updatecheck/updatecheck_test.go]: @rfay the function being tested is ran against version.DdevVersion, so this array is meant to be an example of possible values of that.

If building off a non-release commit in a workspace with uncommitted modifications, I get tags similar to `v0.3-10-gd30045a-dirty`

File: pkg/updatecheck/updatecheck_test.go (hunk 1 above)
Comment (beeradb) [inline pkg/updatecheck/updatecheck_test.go]: This is accurate according to https://golang.org/pkg/time/#Duration

The sleep could possibly be avoided by just altering the modified time.

File: cmd/ddev/cmd/root.go [hunk 5]
```diff
@@ -40,6 +47,31 @@ var RootCmd = &cobra.Command{
...
 			}
 		}
+
+		usr, err := homedir.Dir()
+		if err != nil {
+			log.Fatal("Could not detect users home directory. Is it set?")
+		}
+
+		updateFile := filepath.Join(usr, ".ddev", ".update")
+		// Do periodic detection of whether ran update is available for drud users.
```
Comment (beeradb) [inline cmd/ddev/cmd/root.go]: s/ran/an actually :) good catch!

File: cmd/ddev/cmd/root.go [hunk 6]
```diff
@@ -40,6 +47,31 @@ var RootCmd = &cobra.Command{
...
+		timeToCheckForUpdates, err := updatecheck.IsUpdateNeeded(updateFile, updateInterval)
+		if err != nil {
+			util.Warning("Could not perform update check")
```
Comment (beeradb) [inline cmd/ddev/cmd/root.go]: I tried outputting the error here and I really didn't think the result added value. Here's any example:

`Could not check for updates: Get https://api.github.com/repos/drud/ddev/releases?page=1: dial tcp: lookup api.github.com on [::1]:53: read udp [::1]:52598->[::1]:53: read: connection refused`

Instead, I updated the error message to be:
`Could not check for updates! this is most often caused by a networking issue`

File: pkg/updatecheck/updatecheck_test.go (hunk 4 above)
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Right, I understood that, and read your statement about the strategy in the OP. Just saying that it's hard to know what a bad one would look like. Probably wouldn't be a tag like we use for pushing containers. All good.

File: pkg/updatecheck/updatecheck_test.go [hunk 7]
```diff
@@ -23,20 +23,20 @@ func TestUpdateNeeded(t *testing.T) {
...
+	assert.False(updateRequired, "Update is not required when the update interval has not been met")
 	assert.NoError(err)
 
 	// Sleep for 2 seconds.
 	time.Sleep(time.Duration(2) * time.Second)
 
 	// Ensure updates are required for a duration lower than the sleep.
 	updateRequired, err = IsUpdateNeeded(updateFile, time.Duration(1)*time.Second)
-	assert.True(updateRequired, "Update is not required when the update file does not exist")
+	assert.True(updateRequired, "Update is equired after the update interval has passed")
```
Comment (rfay) [inline pkg/updatecheck/updatecheck_test.go]: Shoot, typo here. Quick tiniest PR for it?

**Pull request merged**